From 6b959ba22d34ca793ffdb15b5715457c78e38b1a Mon Sep 17 00:00:00 2001
From: Yang Jihong <yangjihong1@huawei.com>
Date: Fri, 2 Sep 2022 16:29:18 +0800
Subject: [PATCH] perf/core: Fix reentry problem in perf_output_read_group()
Git-commit: 6b959ba22d34ca793ffdb15b5715457c78e38b1a
Patch-mainline: v6.0
References: bsn#12
Modified-by-SEL: Yes, modified per openEuler commit 1fe6ef1d


perf_output_read_group may respond to IPI request of other cores and invoke
__perf_install_in_context function. As a result, hwc configuration is modified.
causing inconsistency and unexpected consequences.

Interrupts are not disabled when perf_output_read_group reads PMU counter.
In this case, IPI request may be received from other cores.
As a result, PMU configuration is modified and an error occurs when
reading PMU counter:

                     CPU0                                         CPU1
                                                      __se_sys_perf_event_open
                                                        perf_install_in_context
  perf_output_read_group                                  smp_call_function_single
    for_each_sibling_event(sub, leader) {                   generic_exec_single
      if ((sub != event) &&                                   remote_function
          (sub->state == PERF_EVENT_STATE_ACTIVE))                    |
  <enter IPI handler: __perf_install_in_context>   <----RAISE IPI-----+
  __perf_install_in_context
    ctx_resched
      event_sched_out
        armpmu_del
          ...
          hwc->idx = -1; // event->hwc.idx is set to -1
  ...
  <exit IPI>
              sub->pmu->read(sub);
                armpmu_read
                  armv8pmu_read_counter
                    armv8pmu_read_hw_counter
                      int idx = event->hw.idx; // idx = -1
                      u64 val = armv8pmu_read_evcntr(idx);
                        u32 counter = ARMV8_IDX_TO_COUNTER(idx); // invalid counter = 30
                        read_pmevcntrn(counter) // undefined instruction

Signed-off-by: Yang Jihong <yangjihong1@huawei.com>
Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
Link: https://lkml.kernel.org/r/20220902082918.179248-1-yangjihong1@huawei.com
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/events/core.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/kernel/events/core.c b/kernel/events/core.c
index 6fcd4177ade6..21fe33ca6327 100644
--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -6779,6 +6779,13 @@ static void perf_output_read_group(struct perf_output_handle *handle,
 	u64 read_format = event->attr.read_format;
 	u64 values[5];
 	int n = 0;
+	unsigned long flags;
+
+	/*
+	 * Disabling interrupts avoids all counter scheduling
+	 * (context switches, timer based rotation and IPIs).
+	 */
+	local_irq_save(flags);
 
 	values[n++] = 1 + leader->nr_siblings;
 
@@ -6811,6 +6818,8 @@ static void perf_output_read_group(struct perf_output_handle *handle,
 
 		__output_copy(handle, values, n * sizeof(u64));
 	}
+
+	local_irq_restore(flags);
 }
 
 #define PERF_FORMAT_TOTAL_TIMES (PERF_FORMAT_TOTAL_TIME_ENABLED|\
-- 
2.33.0

